### PR TITLE
feat: Implement `.toReadable()` for streams

### PR DESCRIPTION
**Pull request by giraugh**
## Completed
- Adds `.toReadable()` to streams with options for both raw streams and objects
- Adds unit tests for `.toReadable()`

Had to do some wacky stuff with tsdoc to get it to look decent in the generated docs - might be a way to improve it potentially if you wanna give it a whirl. Not sure if there's some way to specifically mark something as an overload.

### CODE REVIEW

File: src/stream/consumption.ts [hunk 1]
```diff
@@ -133,4 +134,92 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+     *
+     * @see {@link Stream#serialize} if the stream values should be serialized to json
+     * @group Consumption
+     */
+    toReadable(kind: "raw" | "object", options?: { single?: boolean; atoms?: boolean }): Readable;
+
+    /**
+     * Produce a readable node stream with the raw values from the stream.
+     *
+     * @param options.single - Whether to emit only the first atom
```
Comment (andogq) [inline src/stream/consumption.ts:155]: Not sure if this single option should be included here. Would it be better to have `.limit` or something chained before?

File: src/stream/consumption.ts [hunk 2]
```diff
@@ -133,4 +134,92 @@ export class StreamConsumption<T, E> extends StreamBase {
 
         return s;
     }
+
+    /**
+     * Produce a readable node stream with the values from the stream.
+     *
+     * @param kind - What kind of readable stream to produce. When "raw" only strings and buffers can be emitted on the stream. Use "object" to preserve
```
Comment (andogq) [inline src/stream/consumption.ts:141]: Not sure how (proper types in node would help), but can we make it clearer that it returns a readable stream that is operating in object mode, when applicable.

File: src/stream/consumption.ts [hunk 3]
```diff
@@ -133,4 +134,92 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+                    );
+                    break;
+                }
+
+                // monitor for non raw values when not using object mode
+                if (
+                    kind === "raw" &&
+                    !(typeof atom.value === "string" || atom.value instanceof Buffer)
+                ) {
+                    s.emit(
```
Comment (andogq) [inline src/stream/consumption.ts]: I think log something out here too

File: src/stream/consumption.ts [hunk 4]
```diff
@@ -133,4 +134,92 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+                    break;
+                }
+
+                // Emit atom or atom value
+                s.push(options?.atoms ? atom : atom.value);
+                sentItems += 1;
+            }
+
+            // End the stream
+            s.push(null);
```
Comment (andogq) [inline src/stream/consumption.ts:216]: Include a note about `null` being in the windpipe stream in the doc comment. Pretty much, if someone has a stream of `[ok('hi'), ok(null), ok('asdf')]`, only 'hi' will actually get emitted on to the stream. Will be an edge-case, maybe even worthwhile including a console.log warning if it's detected...

File: src/stream/consumption.ts [hunk 5]
```diff
@@ -133,4 +134,92 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+        })();
+
+        return s;
```
Comment (andogq) [inline src/stream/consumption.ts:219]: nice work on the iife

File: test/consumption.test.ts [hunk 6]
```diff
@@ -71,4 +72,79 @@ describe.concurrent("stream consumption", () => {
...
+        });
+
+        test("single raw value", async ({ expect }) => {
+            expect.assertions(2);
+
+            const stream = $.from(["hello", " ", "world"]).toReadable("raw", { single: true });
+
+            expect(stream).to.be.instanceof(Readable);
+            const values = await promisifyStream(stream);
+            expect(values.join("")).to.equal("hello");
```
Comment (andogq) [inline test/consumption.test.ts]: include a test case for null being included on the windpipe stream

**Review by andogq — CHANGES_REQUESTED** [review]
Great work, bit of a hairy one! There's just an edge case to consider, as well as some small clarifications to the documentation.

As for the `single` option, I'm suggesting that it's not included for this option. The reason it's required for the `serialise` handler is because we need to know if we wrap the resulting string in `[]` or not. I'd opt that we include a .limit method if it's required to enforce only emitting one item.

File: src/stream/consumption.ts [hunk 7]
```diff
@@ -133,4 +134,86 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+     * @note When not using `options.atoms`, any `null` atom values will be skipped when piping to the readable stream
+     * @see {@link Stream#serialize} if the stream values should be serialized to json
+     * @group Consumption
+     */
+    toReadable(kind: "object", options?: { atoms?: boolean }): Readable;
+
+    toReadable(
+        kind: "raw" | "object",
+        options: { single?: boolean; atoms?: boolean } = {},
+    ): Readable {
```
Comment (andogq) [inline src/stream/consumption.ts]:
```suggestion
    /**
     * Produce a readable node stream with the raw values from the stream.
     *
     * @see {@link Stream#serialize} if the stream values should be serialized to json
     * @group Consumption
     */
    toReadable(kind: "raw"): Readable;

    /**
     * Produce a readable node stream in object mode with the values from the stream
     *
     * @param options.atoms - By default, only `ok` values are emitted, however enabling this
     * will emit all values.
     *
     * @note When not using `options.atoms`, any `null` atom values will be skipped when piping to the readable stream
     * @see {@link Stream#serialize} if the stream values should be serialized to json
     * @group Consumption
     */
    toReadable(kind: "object", options?: { atoms?: boolean }): Readable;
    
    /**
     * Produce a readable node stream with the values from the stream.
     *
     * @param kind - What kind of readable stream to produce. When "raw" only strings and buffers can be emitted on the stream. Use "object" to preserve
     * objects in the readable stream. Note that object values are not serialised, they are emitted as objects.
     * @param options - Options for configuring how atoms are output on the stream
     *
     * @see {@link Stream#serialize} if the stream values should be serialized to json
     * @group Consumption
     */
    toReadable(
        kind: "raw" | "object",
        options?: { atoms?: boolean },
    ): Readable {
```

Just a fix up to the overloaded method signature, no need for it to be duplicated. Run it through your LSP to make sure I didn't break anything.

File: src/stream/consumption.ts [hunk 8]
```diff
@@ -133,4 +134,86 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+            read() {},
+            objectMode: kind === "object",
+        });
+
+        // Spin off asynchronously so that the stream can be immediately returned
+        (async () => {
+            for await (const atom of this) {
+                // Determine whether non-ok values should be filtered out
+                if (options?.atoms !== true && !isOk(atom)) {
+                    continue;
```
Comment (andogq) [inline src/stream/consumption.ts:189]: Doesn't have to be decided for this, but how do you think non-ok atoms should be handled if they're gonna be dropped. Gut instinct wants to log out unknown errors so they're not lost to the void.

This is a bigger issue in Windpipe, so just something to think about.

File: src/stream/consumption.ts [hunk 9]
```diff
@@ -133,4 +134,86 @@ export class StreamConsumption<T, E> extends StreamBase {
...
+
+                // monitor for non raw values when not using object mode
+                if (
+                    kind === "raw" &&
+                    !(typeof atom.value === "string" || atom.value instanceof Buffer)
+                ) {
+                    const message = `Stream indicated it would emit raw values but emitted a '${typeof atom.value}' object`;
+                    console.error(message);
+                    s.emit("error", new Error(message));
+                    break;
```
Comment (andogq) [inline src/stream/consumption.ts:200]: Just to confirm, the stream will be stopped early if an object is detected in raw mode? If so, maybe this could be included in the doc comment.

File: test/creation.test.ts
Comment (andogq) [inline test/creation.test.ts:1]: Nice addition, thanks for catching this.

**Review by andogq — CHANGES_REQUESTED** [review]

File: src/stream/consumption.ts (hunk 7 above)
Comment (giraugh) [reply src/stream/consumption.ts]: I think this was the change I made for the generated docs to work properly

Comment (giraugh) [reply src/stream/consumption.ts]: Yeah, with that setup it uses the "raw" version as the base overload for the purpose of LSP descriptions and in the generated docs. So for example, this is the first mention of `.toReadable()` in the docs
image

**Review by andogq — APPROVED** [review]
